### PR TITLE
Eliminate need for `~` in `~Foo::bar()`

### PR DESCRIPTION
**Pull request by Centril**
Fixes https://github.com/FuelLabs/sway/issues/1436.

This PR first removes the need for `~` in `~Foo::bar` by first delaying the interpretation of `Foo::Bar` (associated call, free function call, or enum variant construction?) until type checking. This is achieved with `ExprKind::AmbiguousPathExpression`. During type checking, we ask whether `Foo` (and any prefixes) is a module, or an enum. If it's neither, we attempt type checking as an associated function call. Otherwise, we continue as if we had a `ExpressionKind::DelineatedPath`.

Once `~` is made redundant, it's then also removed from the language and so also from tests, examples, and the standard library.

### CODE REVIEW

File: sway-error/src/convert_parse_tree_error.rs [hunk 1]
```diff
@@ -9,8 +9,6 @@ pub enum ConvertParseTreeError {
     FunctionArbitraryExpression { span: Span },
     #[error("generics are not supported here")]
     GenericsNotSupportedHere { span: Span },
-    #[error("fully qualified paths are not supported here")]
-    FullyQualifiedPathsNotSupportedHere { span: Span },
```
Comment (mohammadfawaz) [inline sway-error/src/convert_parse_tree_error.rs:13]: Does this mean I can now write
```
    fn test_function(a: lib1::Foo, b: lib2::Foo) -> bool;
```
? That is, can i use call paths when referring to a type?

**Review by mohammadfawaz — COMMENTED** [review]
🥳

File: sway-error/src/convert_parse_tree_error.rs (hunk 1 above)
Comment (mohammadfawaz) [reply sway-error/src/convert_parse_tree_error.rs:13]: Oh this seems related https://github.com/FuelLabs/sway/issues/1644

Comment (mohammadfawaz) [reply sway-error/src/convert_parse_tree_error.rs:13]: Yeah nevermind, I was thinking of a different error here.

**Review by mohammadfawaz — APPROVED** [review]
🚀

**Review by sezna — APPROVED** [review]

**Pull request merged**